File: archive_python/bench.py
"""Linux/MaixPy bench test, NOT E907 firmware. No inferred angle feedback."""
import argparse
import csv
import json
import math
from pathlib import Path
import time

FIELDS = ['seq', 'event', 'servo', 'pwm', 'pulse_us', 'frequency_hz',
          'before_ns', 'after_ns', 'measured_angle_deg', 'status']

def validate(c):
    f = c['frequency_hz']
    if f not in (50, 333):
        raise ValueError('frequency_hz must be 50 or 333')
    if not 0.5 <= c['hold_s'] <= 30 or not 1 <= c['repetitions'] <= 10:
        raise ValueError('invalid hold/repetitions')
    if not c['offsets_us'] or len(c['offsets_us']) > 100:
        raise ValueError('invalid offsets')
    if len(c['channels']) != 4:
        raise ValueError('exactly four channels required')
    mappings = [('B2',4), ('B3',5), ('A30',6), ('A31',7)]
    if sorted((x['pin'],x['pwm']) for x in c['channels']) != sorted(mappings):
        raise ValueError('expected B2/PWM4 B3/PWM5 A30/PWM6 A31/PWM7')
    if sorted(x['servo'] for x in c['channels']) != [1,2,3,4]:
        raise ValueError('servo IDs must be 1..4')
    for x in c['channels']:
        if not 500 <= x['min_us'] <= x['center_us'] <= x['max_us'] <= 2500:
            raise ValueError('invalid pulse bounds')
        for d in [0] + c['offsets_us']:
            p = x['center_us'] + d
            if not math.isfinite(p) or not x['min_us'] <= p <= x['max_us'] or p >= 1e6/f:
                raise ValueError('pulse outside configured limits')

def plan(c):
    for rep in range(c['repetitions']):
        for x in c['channels']:
            for d in c['offsets_us']:
                yield x, x['center_us'] + d

def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument('--config', default=str(Path(__file__).with_name('config.json')))
    ap.add_argument('--out', required=True, help='new run directory; never overwrites')
    ap.add_argument('--run', action='store_true', help='actually enable MaixCAM2 PWM')
    args = ap.parse_args()
    c = json.loads(Path(args.config).read_text(encoding='utf-8'))
    validate(c)
    out = Path(args.out); out.mkdir(parents=True, exist_ok=False)
    c.update(mode='hardware_linux' if args.run else 'DRY_RUN', unix_start=time.time(),
             clock='time.monotonic_ns; command timestamps, not physical motion',
             measured_angle_source='video annotation; absent in commands.csv')
    (out/'metadata.json').write_text(json.dumps(c, ensure_ascii=False, indent=2),encoding='utf-8')
    devices = {}
    old_functions = {}
    seq = 0
    with (out/'commands.csv').open('w',newline='',encoding='utf-8') as stream:
        writer = csv.DictWriter(stream, fieldnames=FIELDS); writer.writeheader()
        def command(x, pulse, event):
            nonlocal seq
            seq += 1; before = time.monotonic_ns()
            status = 'DRY_RUN'
            try:
                if args.run:
                    devices[x['servo']].duty(pulse*c['frequency_hz']/10000.0)
                    status = 'API_OK_NOT_FEEDBACK'
            except Exception:
                status = 'API_ERROR'; raise
            finally:
                after = time.monotonic_ns()
                writer.writerow(dict(seq=seq,event=event,servo=x['servo'],pwm=x['pwm'],
                    pulse_us=pulse,frequency_hz=c['frequency_hz'],before_ns=before,
                    after_ns=after,measured_angle_deg='',status=status)); stream.flush()
                print(seq,event,'servo',x['servo'],'pulse_us',pulse,status,flush=True)
        try:
            if args.run:
                from maix import pwm, pinmap, err, sys as maixsys
                if maixsys.device_id() != 'maixcam2':
                    raise RuntimeError('This pin map is for MaixCAM2 only')
                for x in c['channels']:
                    old_functions[x['pin']] = pinmap.get_pin_function(x['pin'])
                    err.check_raise(pinmap.set_pin_function(x['pin'], 'PWM'+str(x['pwm'])), 'pinmux')
                    dev = pwm.PWM(x['pwm'],freq=c['frequency_hz'],duty=0,enable=False)
                    devices[x['servo']] = dev
                    command(x,x['center_us'],'initial_neutral')
                    err.check_raise(dev.enable(), 'enable PWM')
                time.sleep(3)
            for x,pulse in plan(c):
                command(x,pulse,'step')
                if args.run: time.sleep(c['hold_s'])
        except KeyboardInterrupt:
            print('Interrupted: disable PWM; no forced return movement.')
        finally:
            # Last normal step is neutral. On error/interrupt disable immediately.
            for sid,dev in devices.items():
                try: err.check_raise(dev.disable(), 'disable PWM')
                except Exception as e: print('DISABLE FAILED',sid,repr(e))
            if args.run and old_functions:
                for pin,func in old_functions.items():
                    try: err.check_raise(pinmap.set_pin_function(pin,func),'restore pinmux')
                    except Exception as e: print('PINMUX RESTORE FAILED',pin,repr(e))
            (out/'end.json').write_text(json.dumps({'unix_end':time.time(),'last_seq':seq}),encoding='utf-8')
    print('Saved',out)

if __name__ == '__main__': main()

File: cpp/src/bench.cpp
#include "common.hpp"
#include <memory>
#ifdef SERVO_MAIXCDK
#include "maix_pwm.hpp"
#include "maix_pinmap.hpp"
#include "maix_sys.hpp"
#endif
struct Hardware{
#ifdef SERVO_MAIXCDK
 std::map<int,std::unique_ptr<maix::peripheral::pwm::PWM>> devices;
 std::map<std::string,std::string> old;
 void init(const Json&x,int hz){using namespace maix::peripheral;std::string pin=x.at("pin");int id=x.at("servo"),p=x.at("pwm");old[pin]=pinmap::get_pin_function(pin);require(int(pinmap::set_pin_function(pin,"PWM"+std::to_string(p)))==0,"pinmux failed");devices[id]=std::make_unique<pwm::PWM>(p,hz,0,false);}
 void set(int id,double us){require(devices.at(id)->duty_val(int(std::llround(us*1000)))>=0,"PWM write failed");}
 void enable(int id){require(int(devices.at(id)->enable())==0,"PWM enable failed");}
 ~Hardware(){for(auto&x:devices)try{if(int(x.second->disable())!=0)std::cerr<<"DISABLE FAILED "<<x.first<<'\n';}catch(...){std::cerr<<"DISABLE exception\n";}devices.clear();for(auto&x:old)try{if(int(maix::peripheral::pinmap::set_pin_function(x.first,x.second))!=0)std::cerr<<"RESTORE FAILED "<<x.first<<'\n';}catch(...){std::cerr<<"RESTORE exception\n";}}
#else
 void init(const Json&,int){throw std::runtime_error("Build with MaixCDK for --run");}void set(int,double){}void enable(int){}
#endif
};
int main(int argc,char**argv){try{
 Args a(argc,argv,{"--config","--out"},{"--run"});Json c=Json::parse(read_file(a.get("--config","config.json")));auto steps=plan(c);bool run=a.has("--run");
#ifndef SERVO_MAIXCDK
 require(!run,"This PC binary is dry-run only. Build MaixCDK project for hardware.");
#else
 if(run)require(maix::sys::device_id()=="maixcam2","Wrong board");
#endif
 fs::path out=a.need("--out");new_run(out);c["mode"]=run?"hardware_cpp":"DRY_RUN";c["clock"]="steady_clock nanoseconds; commands NOT motion feedback";write_file(out/"metadata.json",c.dump(2));
 std::signal(SIGINT,on_signal);std::signal(SIGTERM,on_signal);unsigned seq=0;std::string outcome="complete";int result=0;
 std::ofstream log((out/"commands.csv").string());log.exceptions(std::ios::failbit|std::ios::badbit);log<<"seq,event,servo,pwm,pulse_us,frequency_hz,before_ns,after_ns,measured_angle_deg,status\n";
 try{Hardware hw;int hz=c.at("frequency_hz");auto command=[&](Step s,const char*event){auto before=now_ns();std::string status=run?"API_OK_NOT_FEEDBACK":"DRY_RUN";bool failed=false;try{if(run)hw.set(s.servo,s.pulse);}catch(...){status="API_ERROR";failed=true;}
 log<<++seq<<','<<event<<','<<s.servo<<','<<s.pwm<<','<<s.pulse<<','<<hz<<','<<before<<','<<now_ns()<<",,"<<status<<'\n';log.flush();std::cout<<seq<<" servo="<<s.servo<<" pulse_us="<<s.pulse<<' '<<status<<std::endl;require(!failed,"PWM command failed");};
 if(run){for(auto&x:c["channels"]){if(interrupted)break;hw.init(x,hz);command({x["servo"],x["pwm"],x["center_us"]},"initial_neutral");hw.enable(x["servo"]);}sleep_interruptible(3000);}
 for(auto&s:steps){if(interrupted)break;command(s,"step");if(run)sleep_interruptible(int(c["hold_s"].get<double>()*1000));}
 if(interrupted){outcome="interrupted";result=130;}
 }catch(const std::exception&e){outcome="error";result=1;std::cerr<<e.what()<<'\n';}
 write_file(out/"end.json",Json{{"last_seq",seq},{"outcome",outcome}}.dump(2));return result;
 }catch(const std::exception&e){std::cerr<<e.what()<<'\n';return 1;}}
